frontend/packages/console-shared/src/hooks/perspective-utils.ts

```diff
@@ -0,0 +1,166 @@
+      obj = perspectiveExtensions.reduce((acc, perspectiveExtension) => {
+        const perspective = perspectives?.find((p) => p.id === perspectiveExtension.properties.id);
+
+        if (
+          !perspective?.visibility?.state ||
+          perspective.visibility.state === PerspectiveVisibilityState.Enabled
+        ) {
+          acc[perspectiveExtension.properties.id] = true;
+        } else if (perspective?.visibility?.state === PerspectiveVisibilityState.Disabled) {
+          acc[perspectiveExtension.properties.id] = false;
+        }
+        return acc;
+      }, {});
+    }
+    return obj;
+  });
+
+  const handleResults = React.useCallback(
+    (id: string, newState: boolean) => {
+      setResults((oldResults: Record<string, boolean>) => {
+        if (oldResults[id] === newState) {
+          return oldResults;
+        }
+        return {
+          ...oldResults,
+          [id]: newState,
+        };
+      });
+    },
+    [setResults],
+  );
+  React.useEffect(() => {
+    if (window.SERVER_FLAGS.perspectives) {
+      const perspectives: Perspective[] = JSON.parse(window.SERVER_FLAGS.perspectives);
+      perspectiveExtensions.forEach((perspectiveExtension) => {
+        const perspective = perspectives?.find((p) => p.id === perspectiveExtension.properties.id);
+
+        if (
+          !perspective ||
+          !perspective.visibility ||
+          perspective.visibility.state === PerspectiveVisibilityState.Enabled
+        ) {
+          handleResults(perspectiveExtension.properties.id, true);
+        } else if (perspective.visibility.state === PerspectiveVisibilityState.Disabled) {
+          handleResults(perspectiveExtension.properties.id, false);
+        } else if (
+          perspective?.visibility?.state === PerspectiveVisibilityState.AccessReview &&
+          perspective?.visibility?.accessReview &&
+          Object.keys(perspective?.visibility?.accessReview)?.length > 0
+        ) {
+          hasReviewAccess(perspective?.visibility?.accessReview)
+            .then((res) => {
+              handleResults(perspectiveExtension.properties.id, res);
+            })
+            .catch((e) => {
+              handleResults(perspectiveExtension.properties.id, true);
+              // eslint-disable-next-line no-console
+              console.warn('Could not check access', e);
+            });
+        }
+      });
+    }
+  }, [perspectiveExtensions, handleResults]);
+  const perspectives = React.useMemo(() => {
+    const filteredExtensions = perspectiveExtensions.filter((e) => results[e.properties.id]);
+
+    if (!window.SERVER_FLAGS.perspectives) {
+      return perspectiveExtensions;
+    }
+
+    return filteredExtensions.length === 0 &&
+      Object.keys(results).length === perspectiveExtensions.length
+      ? perspectiveExtensions.filter((p) => p.properties.id === 'admin')
+      : filteredExtensions;
+  }, [perspectiveExtensions, results]);
```

Comment:
The return could be before the filter, but this is just a nit and we can go forward anyway.

```suggestion
  const perspectives = React.useMemo(() => {
    if (!window.SERVER_FLAGS.perspectives) {
      return perspectiveExtensions;
    }

    const filteredExtensions = perspectiveExtensions.filter((e) => results[e.properties.id]);
    return filteredExtensions.length === 0 &&
      Object.keys(results).length === perspectiveExtensions.length
      ? perspectiveExtensions.filter((p) => p.properties.id === 'admin')
      : filteredExtensions;
  }, [perspectiveExtensions, results]);
```